### PR TITLE
Hibernate projection inside Hibernate with Panache

### LINKED ISSUE #6261: Provides query projection for Hibernate with Panache
**Description**
Provides the same projection capabilities than for MongoDB with Panace.

See PR #4202

### PR DESCRIPTION
**Pull request by loicmathieu**
This implements the `PanacheQuery.project(Class)` method that allows doing a DTO projection on the query.

The projection class needs to be a Java Bean and have a constructor with all it's fields in it. In order to work in native it needs to be annotated with `@RegisterForReflection`.

It uses the **constructor expression** construct, to build it we need to have the compiler parameters attribute on (by default on our Quarkus Maven archetype but not on the IT).

It fixes #6261

### CODE REVIEW

**Review by gsmet — DISMISSED** [review]
Juste a note that we will need a squash before merging. Feel free to dismiss as soon as it's done.

File: docs/src/main/asciidoc/hibernate-orm-panache.adoc
Comment (FroMage) [inline docs/src/main/asciidoc/hibernate-orm-panache.adoc:587]:
```suggestion
This is also called **dynamic instantiation** or **constructor expression**, more info can be found on the Hibernate guide:
```

File: docs/src/main/asciidoc/hibernate-orm-panache.adoc
Comment (FroMage) [inline docs/src/main/asciidoc/hibernate-orm-panache.adoc:591]: Why? What happens if there are more?

File: docs/src/main/asciidoc/hibernate-orm-panache.adoc
Comment (FroMage) [inline docs/src/main/asciidoc/hibernate-orm-panache.adoc:599]: Can't we turn this into a public final field?

File: docs/src/main/asciidoc/hibernate-orm-panache.adoc
Comment (FroMage) [inline docs/src/main/asciidoc/hibernate-orm-panache.adoc:623]:
```suggestion
The implementation of the `project(Class)` method uses the constructor's parameter names to build the select clause of the query,
```

File: docs/src/main/asciidoc/hibernate-orm-panache.adoc
Comment (FroMage) [inline docs/src/main/asciidoc/hibernate-orm-panache.adoc:625]:
```suggestion
This is enabled by default if you are using the Quarkus Maven archetype. If you are not using it, add the property `<maven.compiler.parameters>true</maven.compiler.parameters>` to your pom.xml.
```

File: docs/src/main/asciidoc/hibernate-orm-panache.adoc
Comment (FroMage) [inline docs/src/main/asciidoc/hibernate-orm-panache.adoc:630]: If it updates the query (mutate) then order doesn't matter. I think what you mean is that it will create a new query. In either case order should not matter because the query copy can use settings from the original query, no?

File: extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/runtime/PanacheQueryImpl.java
Comment (FroMage) [inline extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/runtime/PanacheQueryImpl.java:101]: Huh, this is a sneaky mutation that changes the type of this query. I'm not sure this is a great idea. Why not return a new query object?

**Review by FroMage — CHANGES_REQUESTED** [review]
This looks good but needs some tweaking.

File: docs/src/main/asciidoc/hibernate-orm-panache.adoc
Comment (loicmathieu) [reply docs/src/main/asciidoc/hibernate-orm-panache.adoc:591]: I took the first one.
If there are more, we will need a way to know which one to choose to build the select clause.
We can do it by adding an annotation to denote the correct constructor, but I prefere to keep it simple for the moment and only allow one constructor.
I can add a check in the code and fail if there is more than one constructor
We can revisit this later if needed.

File: docs/src/main/asciidoc/hibernate-orm-panache.adoc
Comment (loicmathieu) [reply docs/src/main/asciidoc/hibernate-orm-panache.adoc:599]: Hibernate documentation said that the class must be a Java Bean.
I just tested it and it can be a POJO with public fields.
I'll update the documentation as this is more consistent with what we do in other areas.

File: extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/runtime/PanacheQueryImpl.java
Comment (loicmathieu) [reply extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/runtime/PanacheQueryImpl.java:101]: PanacheQuery is a statefull object where you can mutate the underlying JPA query object via multiple operations (lock, hint, pagination, ...).
So I stick to the same principle here, projection is just another way to mutate the query before firing it to the database by adding a SELECT clause.

I start thinking that a refactoring should be needed to implements something closer to a builder, each operation on PanacheQuery prior to retrieving the list of result should just update PanacheQuery internal state then when we get the resultList we build the JPA query.

With this, the order of the operation done on the query (see your previous comment) has no importance.

File: docs/src/main/asciidoc/hibernate-orm-panache.adoc
Comment (loicmathieu) [reply docs/src/main/asciidoc/hibernate-orm-panache.adoc:630]: See my next comment. I think some refactoring needs to be done to mitigate this.

File: extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/runtime/PanacheQueryImpl.java
Comment (loicmathieu) [reply extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/runtime/PanacheQueryImpl.java:101]: And this is what we did for MongoDB with Panache: we create a projection document and change the type of the query. So it's consistent!
https://github.com/quarkusio/quarkus/blob/master/extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/runtime/PanacheQueryImpl.java#L45

Comment (FroMage) [reply extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/runtime/PanacheQueryImpl.java:101]: Well let me show you the problem with changing the runtime type of an object:

```java
PanacheQuery<Order> query = Order.find("name", "Stef");
PanacheQuery<OrderName> otherQuery = query.project(OrderName.class);
Order order = query.firstResult(); // ClassCastException
```

It's fine to mutate the state of an object, but mutating its type is generally not done.

File: docs/src/main/asciidoc/hibernate-orm-panache.adoc
Comment (FroMage) [reply docs/src/main/asciidoc/hibernate-orm-panache.adoc]: Cool

File: docs/src/main/asciidoc/hibernate-orm-panache.adoc
Comment (FroMage) [reply docs/src/main/asciidoc/hibernate-orm-panache.adoc:591]: OK fine.

File: extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/runtime/PanacheQueryImpl.java
Comment (loicmathieu) [reply extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/runtime/PanacheQueryImpl.java:101]: I agree this is not convenient but if we create a new `PanacheQuery` for projection it will complexify the implementation and adds maintenance burden in case we want to copy the state of the previous query to the next. Each time we update `PanacheQuery` state management we will need to take care of updating the project method ...

The parameter type of `PanacheQuery` is the return type of the query, this is a component of it's state but I agree this is not very elegant.

I'll take some time to think about this ;)

Comment (FroMage) [reply extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/runtime/PanacheQueryImpl.java:101]: Just make a constructor of `PanacheQueryImpl(PanacheQueryImpl)` which copies the setup and it's all in one place when we add new state. Don't do that in the `project` method.

Comment (loicmathieu) [reply extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/runtime/PanacheQueryImpl.java]: Good idea, I implemented it

**Comment by loicmathieu** [comment]
@FroMage I implement the creation of a new PanacheQuery in case of projection.
I also refactor PanacheQueryImpl to create the JPA query lazilly.
I also fix a bug for the creation of the count query that remove the order by clause, and make sure it will works for projection based query.

If it's OK for you, I'll squash the commits.
I'll also create an issue for MongoDB with Panache to refactor the project method to do the same: create a new PanacheQuery.

**Comment by FroMage** [comment]
Could you rebase please there's a conflict.

**Comment by loicmathieu** [comment]
@FroMage rebase done.

File: extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/runtime/PanacheQueryImpl.java
Comment (FroMage) [inline extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/runtime/PanacheQueryImpl.java:318]: I think this method should be called `createQueryNoPaging` or something like that.

File: extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/runtime/PanacheQueryImpl.java
Comment (FroMage) [inline extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/runtime/PanacheQueryImpl.java:33]: this name feels wrong, isn't it an order-by clause? How about `orderByPart`?

File: integration-tests/hibernate-orm-panache/src/main/java/io/quarkus/it/panache/TestEndpoint.java
Comment (FroMage) [inline integration-tests/hibernate-orm-panache/src/main/java/io/quarkus/it/panache/TestEndpoint.java:1089]: Can we add some tests with queries and parameters too please?

**Review by FroMage — CHANGES_REQUESTED** [review]
Thanks, I prefer the new approach, but I've noticed a few issues. This is really close now :)

File: extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/runtime/PanacheQueryImpl.java
Comment (loicmathieu) [reply extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/runtime/PanacheQueryImpl.java:33]: Yes simply `orderBy` feels better

File: extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/runtime/PanacheQueryImpl.java
Comment (loicmathieu) [reply extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/runtime/PanacheQueryImpl.java:318]: `createBaseQuery` reads better no ?

File: integration-tests/hibernate-orm-panache/src/main/java/io/quarkus/it/panache/TestEndpoint.java
Comment (loicmathieu) [reply integration-tests/hibernate-orm-panache/src/main/java/io/quarkus/it/panache/TestEndpoint.java:1089]: I agree it misses some tests :)

**Comment by loicmathieu** [comment]
@FroMage I rename what needs to be renamed and add some tests (I can add more if needed).
I also squashed the commits so it's ready to be merged.

**Review by FroMage — APPROVED** [review]
Great, thanks!

**Comment by FroMage** [comment]
But you have test failures :(

**Comment by loicmathieu** [comment]
@FroMage sorry about that, I don't understand how it have slip throught.
Test fixed now, CU should become green soon.

**Comment by FroMage** [comment]
Well, squash it, then ;)

**Comment by loicmathieu** [comment]
> Well, squash it, then ;)
Already did it.

**Comment by FroMage** [comment]
Great.

**Comment by FroMage** [comment]
@gsmet can you dismiss your review then?

**Comment by loicmathieu** [comment]
@FroMage gsmet said that we can dismiss its review but I don't have the right to do it in github. Do you ?

**Comment by FroMage** [comment]
Done.

**Comment by emmanuelbernard** [comment]
Sorry for arriving after the battle.
The solution is simple enough. Quickly glimpsing at it, I had two remarks
1. it's only constructor based DTOs, I personally prefer properties but I guess I can live with it and see user feedback
2. the `@RegisterForReflection` being mandatory really sucks. I wish we had a natural way to discover the class. Thoughts?

**Comment by loicmathieu** [comment]
@emmanuelbernard it's after but not too late, it will be in 1.5 so we can still change some bits.

> it's only constructor based DTOs, I personally prefer properties but I guess I can live with it and see user feedback

We didn't come up with a better solution based on what JPA/Hibernate offers. We didn't want to use a mapper between the entity class and the projection class so using Tuple or ResultTransformer was not a best option.

> the @RegisterForReflection being mandatory really sucks. I wish we had a natural way to discover the class. Thoughts?

Yes, in MondoDB we have the `@ProjectionFor` annotation (used to map the projection DTO to the field of the database automatically), so we register all classes annotated with it. 
I don't know if we can retrieve via Jandex all the callsite of a method and access its parameters (and I don't know if it will not be a performance issue).
If this projection DTO is used somewhere we already register classes for reflection (a resteasy ressource operation return type for example) this will not be an issue.

**Comment by emmanuelbernard** [comment]
No Jandex does not record call site usages.

My worry is more that the DTO class is used in the application directly but not the constructor which will be dead code eliminated. Won't be an easy bug to track.

**Pull request merged**
